### PR TITLE
[inductor] Fuse interleaved epilogues in nested reductions

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #191974
* #190596
* #191775
* __->__ #190595

Human context: 

We had previous nested reductions, as well as nvfp4 sub-iteration domain fusing. this combines the two.

Extend fused nested reductions with a sub-parent domain for interleaved pair consumers. This keeps RMSNorm plus NVFP4/MXFP4 packing in one kernel while validating source layouts, non-leaf consumers, and ordinary scheduler fusion gates.

Represent the sub-parent domain in grouped coordinates `(outer, groups, local/factor)`, matching the derived codegen family. This intentionally rejects reshapes that cross grouped-axis boundaries even when their flattened numel matches. Only nested reduced outputs are marked for forced broadcast materialization; the standalone path continues to rely on planned sources and removed-buffer handling.

Test Plan:

```
python test/inductor/test_inductor_scheduler.py -k nested_reduction_sub_parent_domain -v
TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python test/inductor/test_nested_reduction.py -k sub_parent
```

Authored with assistance from Codex.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: test/inductor/test_nested_reduction.py [hunk 1]
```diff
@@ -757,6 +765,202 @@ def check_reduction_fusion(nodes):
             self.check_numeric(f, (x, w))
         self.check_fusion(1 if expect_fullres_consumer else None)
 
+    @inductor_config.patch({"combo_kernels": True, "emulate_precision_casts": True})
+    def test_combo_kernels_skip_nested_reductions(self):
+        import torch.nn.functional as F
```
Comment (drisspg) [inline test/inductor/test_nested_reduction.py:807]: ditto on do we have any correctenss test?

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -1112,6 +1133,14 @@ def _classify_grouped_pointwise_nodes(
                 domain = cls.PointwiseDomain.REDUCED
             elif V.graph.sizevars.statically_known_equals(sn_numel, full_numel):
                 domain = full_domain
+            elif (
+                is_consumer
+                and has_parent_half_domain
+                and V.graph.sizevars.statically_known_equals(
+                    cls.PARENT_HALF_FACTOR * sn_numel, full_numel
```
Comment (drisspg) [inline torch/_inductor/scheduler.py]: Codex comment: 

What happens when `group_size == 2`? In that case the reduced and `PARENT_HALF` domains have the same numel, so the earlier `REDUCED` check always wins and pair consumers skip the interleaved source materialization. Could we either disambiguate this from the source indexing/layout or reject parent-half fusion for `G=2`, and add a `G=2` regression alongside `test_producer_consumer_rmsnorm_interleaved_pair_epilogue`?

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -571,6 +571,10 @@ class PointwiseDomain(enum.Enum):
         LOCAL_REDUCTION_INPUT = enum.auto()
         # Outer reduction tile after broadcast-back, e.g. [B, D].
         PARENT_FULL = enum.auto()
+        # Parent tile with the grouped axis packed two lanes at a time.
```
Comment (drisspg) [inline torch/_inductor/scheduler.py]: # Parent tile with pairs packed along the grouped axis, changing [..., R] to [..., R // 2]. ?

File: torch/_inductor/scheduler.py [hunk 4]
```diff
@@ -1079,6 +1099,7 @@ def _classify_grouped_pointwise_nodes(
         full_numel = V.graph.sizevars.simplify(
             domain_context.grouped_numel * domain_context.grouped_rnumel
         )
+        has_parent_half_domain = domain_context.parent_half_domain is not None
```
Comment (drisspg) [inline torch/_inductor/scheduler.py]: maybe inline?

File: torch/_inductor/scheduler.py [hunk 5]
```diff
@@ -8326,7 +8492,9 @@ def _can_fuse(
         if isinstance(node1, FusedNestedReductions):
             return node1.can_fuse_with(node2, can_reorder=can_reorder)
         if isinstance(node2, FusedNestedReductions):
-            return False
+            # Candidate scoring can present a downstream consumer first.
+            # can_fuse_with still verifies it really consumes the grouped stage.
+            return node2.can_fuse_with(node1, can_reorder=can_reorder)
```
Comment (drisspg) [inline torch/_inductor/scheduler.py]: CODEX: 
Are we sure reversing the arguments here preserves the same dependency and reordering checks as the normal fusion direction? `can_fuse_with()` checks that `node1` consumes the nested stage, but could we add a regression where candidate scoring presents the pair in both orders and verifies the same legal fusion?

**Review by drisspg — APPROVED** [review]
I think the groupsize 2 thing though might be real so have a look :)

**Comment by drisspg** [comment]
@claude Please do a deep review focused on subtle correctness bugs and regressions that are easy for human reviewers to miss. Prioritize findings over summary. I dont care about anything that isnt a real bug drop all the noise but that, if you there isnt anything to report just say so with a small comment

**Comment by yyetim** [comment]
I didn't see any AMD tests in the files (might have missed). Could add some tests for MI350 if infra is available?
